alloc-fmt/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+
+#[doc(hidden)]
+pub static STDERR_MTX: spin::Mutex<()> = spin::Mutex::new(());
+#[doc(hidden)]
+pub static STDOUT_MTX: spin::Mutex<()> = spin::Mutex::new(());
+
+#[doc(hidden)]
+pub static STDOUT: libc::c_int = 1;
+#[doc(hidden)]
+pub static STDERR: libc::c_int = 2;
+
+#[doc(hidden)]
+pub struct FDWriter(pub libc::c_int);
+
+impl Write for FDWriter {
+    #[inline]
+    fn write_str(&mut self, s: &str) -> Result {
+        let mut buf = s.as_bytes();
+        while !buf.is_empty() {
+            unsafe {
+                let written = libc::write(self.0, buf.as_ptr() as *const _, buf.len());
+                if written < 1 {
+                    core::intrinsics::abort();
+                }
+                buf = &buf[written as usize..];
+            }
+        }
+        Ok(())
+    }
+}
+
+// We can't simply 'pub use core::intrinsics::abort' because its use requires
+// feature(core_intrinsics), which the user would have to enable.
+#[doc(hidden)]
+pub unsafe fn abort() -> ! {
+    core::intrinsics::abort();
+}
+
+#[doc(hidden)]
+#[macro_export]
+macro_rules! print_internal {
+    ($file:expr, $mtx:expr, $($arg:tt)*) => {
+        // in case we're called from inside an unsafe block
+        #[allow(unused_unsafe)]
+        unsafe {
+            let guard = $mtx.lock();
+
+            let mut fd = $crate::FDWriter($file);
+            use $crate::Write;
+            let _ = write!(&mut fd, $($arg)*).map_err(|_| {
+                $crate::abort();
+            });
+
+            // explicitly drop the guard so we're guaranteed it lives at least this long;
+            // since we don't actually use the guard, the compiler could theoretically
+            // drop it sooner.
+            $crate::drop(guard);
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! alloc_print {
+    ($($arg:tt)*) => (print_internal!($crate::STDOUT, $crate::STDOUT_MTX, $($arg)*))
+}
+
+#[macro_export]
+macro_rules! alloc_eprint {
+    ($($arg:tt)*) => (print_internal!($crate::STDERR, $crate::STDERR_MTX, $($arg)*))
+}
+
+#[macro_export]
+macro_rules! alloc_println {
+    () => (alloc_print!("\n"));
+    ($fmt:expr) => (alloc_print!(concat!($fmt, "\n")));
+    ($fmt:expr, $($arg:tt)*) => (alloc_print!(concat!($fmt, "\n"), $($arg)*));
+}
+
+#[macro_export]
+macro_rules! alloc_eprintln {
+    () => (alloc_eprint!("\n"));
+    ($fmt:expr) => (alloc_eprint!(concat!($fmt, "\n")));
+    ($fmt:expr, $($arg:tt)*) => (alloc_eprint!(concat!($fmt, "\n"), $($arg)*));
+}
+
+// Sometimes backtraces allocate. If that allocation causes an assertion failure, then we can get
+// into an infinite recursion scenario. In order to detect this, we set this global bool to true,
+// and load it before printing a backtrace. If we find that it is true, we immediately abort
+// (essentially short-circuiting what would eventually happen if print_backtrace_and_abort
+// successfully finished executing).
+#[doc(hidden)]
+pub static IS_ASSERTING: AtomicBool = ATOMIC_BOOL_INIT;
```

Comment:
Potential TODO:

have a static array of first N prime numbers, make this an AtomicUsize and have each thread multiply IS_ASSERTING by primes[thread::current().id() % primes.size()] and divide when finished. The check then asks when IS_ASSERTING is divisible by this prime; cuts down on false positives.